query/src/main/java/org/infinispan/query/impl/massindex/MassIndexerProgressState.java

```diff
@@ -0,0 +1,40 @@
+package org.infinispan.query.impl.massindex;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.hibernate.search.util.common.impl.Futures;
+
+import org.infinispan.query.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+public class MassIndexerProgressState {
+
+   private static final Log LOG = LogFactory.getLog(IndexUpdater.class, Log.class);
+
+   private final MassIndexerProgressNotifier notifier;
+   private final CompletableFuture<?> lastFuture = CompletableFuture.completedFuture( null );
+
+   public MassIndexerProgressState(MassIndexerProgressNotifier notifier) {
+      this.notifier = notifier;
+   }
+
+   public void addItem(Object key, Object value, CompletableFuture<?> future) {
+      // This is what HS5 currently does, but introduce chunking could be a good idea.
+      future.whenComplete((result, exception) -> {
+         if (exception != null) {
+            notifier.notifyEntityIndexingFailure(value.getClass(), key, exception);
+         } else {
+            notifier.notifyDocumentsAdded(1);
+         }
+      }).thenCombine(lastFuture, (ignored1, ignored2) -> null);
```

Comment:
You forgot to update `lastFuture`.

Reply:
Please also check whether this method is called concurrently; if so, updating `lastFuture` will be tricky.

Reply:
Remove the final and update the last future with the result of the combine.
I don't think an instance of `MassIndexerProgressState` is used by more threads concurrently.
`MassIndexerProgressState` is a method local variable and it is used only within its method execution...

Reply:
Anyway I'm not sure that the combine two by two solution perform better then the`CompletableFuture.allOf`. The latter maybe looks less blocking...